server/autolinkplugin/plugin.go

```diff
@@ -61,7 +61,29 @@ func contains(team string, channel string, list []string) bool {
 	return false
 }
 
+func (p *Plugin) isBotUser(userId string) (bool, *model.AppError) {
+	user, appErr := p.API.GetUser(userId)
+	if appErr != nil {
+		p.API.LogError("failed to check if message for rewriting was send by a bot", "error", appErr)
+		return false, appErr
+	}
+
+	return user.IsBot, nil
+}
+
 func (p *Plugin) ProcessPost(c *plugin.Context, post *model.Post) (*model.Post, string) {
+	isBot, appErr := p.isBotUser(post.UserId)
```

Comment:
It is likely considerably more expensive to go back to the server to get the user record than first evaluating to see if any links apply. 3/5 we should move this check after an applicable link is found.

Reply:
@levb I am curious - what does 3/5, 0/5 mean? I often see it in reviews.

Reply:
Wrt. your comment - we do not really have data on how big/how numerous the rules can get, and how much traffic we may need to rewrite and how the plugin system and querying the server performs under load. The rewriting rules also already call `p.API.GetChannel` and `p.API.GetTeam` api calls which may be heavier than the `p.API.GetUser` that we are adding and actually having `isBot` check early may save CPU cycles.

I believe optimizing it now is a bit early. Would you mind if we ask @hanzei to add his two cents?

Reply:
@levb @hanzei Ping.

Reply:
(waiting for @hanzei also cc @lieut-data) A single RPC/post is too expensive, anywhere in a plugin, period. We should move the RPCs to after the match, so we only execute them when needed.

Now, I see the argument that if this PR follows the existing pattern, we should stick with it as is, and have a separate ticket to optimize. I'm 1/5 that we shouldn't - let's at least make sure this PR doesn't add overhead.

https://docs.mattermost.com/process/training.html#id8

Reply:
@levb, based on @ali-farooq0's preliminary analysis of the RPC overhead on community, it /seems/ like RPC, itself, isn't that expensive. The real cost is borne by the database queries to resolve this information, but I'm assuming we're already doing one or more of those to post the message in the first place.

That being said, no objections to moving this check below and wrapping in a `post.Message != postText` test first. Agree we should have a ticket to optimize regardless.

Reply:
Thanks for the discussion! Requested changes can be found in ` Do not rewrite messgaes from bots - review fixes #1` commit.